### PR TITLE
SAC multidimensional actions

### PR DESCRIPTION
**Pull request by albheim**
Adds support for multidimensional actions in SAC. 

Started branch based on the one for #172 since I wanted to build on that, not sure if that was wrong since now all those commits also show up here.

### CODE REVIEW

**Review by findmyway — APPROVED** [review]
LGTM. But we'd better to merge that dependency first and then this one.

File: src/algorithms/policy_gradient/sac.jl [hunk 1]
```diff
@@ -99,12 +99,12 @@ function (p::SACPolicy)(env)
         s = state(env)
         s = Flux.unsqueeze(s, ndims(s) + 1)
         # trainmode:
-        action = evaluate(p, s)[1][] # returns action as scalar
+        action = dropdims(evaluate(p, s)[1], dims=2) # Single action vec, drop second dim
```
Comment (albheim) [inline src/algorithms/policy_gradient/sac.jl:109]: With this the action will always be returned as a vector. I'm not sure how you would prefer to have it, if scalars should be treated special in some way. But for now this seemed clean and easy to me so I went this way.

File: src/experiments/rl_envs/JuliaRL_SAC_Pendulum.jl [hunk 2]
```diff
@@ -58,15 +59,15 @@ function RLCore.Experiment(
...
+            start_policy = RandomPolicy(Space([-1.0..1.0 for _ in 1:na]); rng = rng),
             update_after = 1000,
             update_every = 1,
             rng = rng,
         ),
         trajectory = CircularArraySARTTrajectory(
             capacity = 10000,
             state = Vector{Float32} => (ns,),
-            action = Float32 => (),
+            action = Float32 => (na,),
```
Comment (albheim) [inline src/experiments/rl_envs/JuliaRL_SAC_Pendulum.jl]: Feels to me like this should be `action = Vector{Float32} => (na,),` but it seems like both work. I'm not really sure how the trajectories are built, will take a look later unless someone has more insight and can tell me.

Comment (findmyway) [reply src/experiments/rl_envs/JuliaRL_SAC_Pendulum.jl]: Yes, they are essentially the same. But  `action = Vector{Float32} => (na,),` is a more standard style to align with the doc.

File: src/algorithms/policy_gradient/sac.jl (hunk 1 above)
Comment (findmyway) [reply src/algorithms/policy_gradient/sac.jl:109]: I agree this way is cleaner. But it's strange to me that the experiment of SAC with pendulum doesn't throw an exception here.

https://github.com/JuliaReinforcementLearning/ReinforcementLearningEnvironments.jl/blob/db5a7929f0683709c1a6f7936ee770bad9e78f17/src/environments/examples/PendulumEnv.jl#L92

Comment (albheim) [reply src/algorithms/policy_gradient/sac.jl:109]: I call it with `x[1]` through the ActionTransformedEnv in the Experiment.
```julia
        action_mapping = x -> low + (x[1] + 1) * 0.5 * (high - low),
```

**Comment by albheim** [comment]
@rbange Do you have any source from which this was implemented? Had a look at the original SAC article and was not that implementation at least. Think it could be good to link to in the docs if possible so one could check it out easily.

File: src/experiments/rl_envs/JuliaRL_SAC_Pendulum.jl [hunk 3]
```diff
@@ -76,9 +77,16 @@ function RLCore.Experiment(
...
+        DoEveryNStep() do t, agent, env
             with_logger(lg) do
-                @info "training" reward = total_reward_per_episode.rewards[end]
+                @info(
+                    "training",
+                    reward_term = agent.policy.reward_term,
+                    entropy_term = agent.policy.entropy_term,
+                )
+                if is_terminated(env)
+                    @info "training" reward = total_reward_per_episode.reward log_step_increment = 0
```
Comment (albheim) [inline src/experiments/rl_envs/JuliaRL_SAC_Pendulum.jl:88]: If I use `total_reward_per_episode.rewards[end]` here it errors since the reward is added to the vector in the PostEpisodeStage while the DoEveryNStep is called in the PostActStage. It works for PPO_Pendulum because there it is using TotalBatchRewardPerEpisode which does both the collecting of rewards and the termination check in the PostActStage. Just seemed a bit odd and inconsistent to me so thought I would note it.

File: src/algorithms/policy_gradient/sac.jl [hunk 4]
```diff
@@ -23,6 +23,9 @@ mutable struct SACPolicy{
     update_every::Int
     step::Int
     rng::R
+    # Logging
+    reward_term::Float32
+    entropy_term::Float32
```
Comment (albheim) [inline src/algorithms/policy_gradient/sac.jl:28]: Added terms for the reward and entropy part of the policy loss to make it easier to track and log those. Thought it could be good to have that access when choosing the weighting factor alpha.

File: src/algorithms/policy_gradient/sac.jl (hunk 1 above)
Comment (findmyway) [reply src/algorithms/policy_gradient/sac.jl:109]: I see, thanks for your explanation.

**Comment by rbange** [comment]
> @rbange Do you have any source from which this was implemented? Had a look at the original SAC article and was not that implementation at least. Think it could be good to link to in the docs if possible so one could check it out easily.

As i found no Julia reference implementation, I created it myself. The algorithm itself has many different versions as the authors are continuously improving the algorithm.

The implementation in this repo is based on the initial paper: 

> Haarnoja, T., Zhou, A., Abbeel, P., & Levine, S. (2018). Soft Actor-Critic: Off-Policy Maximum Entropy Deep Reinforcement Learning with a Stochastic Actor. ArXiv:1801.01290 [Cs, Stat]. http://arxiv.org/abs/1801.01290

But without a value function as the authors later proved that such function did not improve results.

Within my thesis i also implemented the `Automating Entropy Adjustment` initially presented in:

> Haarnoja, T., Zhou, A., Hartikainen, K., Tucker, G., Ha, S., Tan, J., Kumar, V., Zhu, H., Gupta, A., Abbeel, P., & Levine, S. (2019). Soft Actor-Critic Algorithms and Applications. ArXiv:1812.05905 [Cs, Stat]. http://arxiv.org/abs/1812.05905

But i haven't found the time yet to adjust the implementation within this repo. Im willing to share that code if you have time to adjust the implementation.

If you want to link a paper i would recommend to link the later one. The latest development state can be tracked in the authors official repository: https://github.com/rail-berkeley/softlearning (They use tensorflow)

**Comment by albheim** [comment]
> The implementation in this repo is based on the initial paper:
> 
> > Haarnoja, T., Zhou, A., Abbeel, P., & Levine, S. (2018). Soft Actor-Critic: Off-Policy Maximum Entropy Deep Reinforcement Learning with a Stochastic Actor. ArXiv:1801.01290 [Cs, Stat]. http://arxiv.org/abs/1801.01290
> 
> But without a value function as the authors later proved that such function did not improve results.

Okay, then I understand that part at least. But I'm still a bit lost on this part from the `evaluate` function
```julia
logp_π = sum(logpdf.(π_dist, z), dims = 1)
logp_π -= sum((2.0f0 .* (log(2.0f0) .- z - softplus.(-2.0f0 * z))), dims = 1)
```
which is not what I find in either of the papers where they have something like
```julia
logp_π = sum(logpdf.(π_dist, z), dims = 1)
logp_π -= sum(log.(1 .- tanh.(z).^2), dims = 1)
```
I seems like it is pretty similar, think I got your version to become 
```julia
logp_π = sum(logpdf.(π_dist, z), dims = 1)
logp_π -= sum(-log(2) .+ log.(1 .- cosh.(z).^2), dims = 1)
```
after some massaging, but I don't understand why one would use that instead of what is presented in the paper.

I might very well have made a mistake or misunderstood something, but something is not immediately obvious with the implementation at least.

EDIT:
Yeah, I had made a mistake. They are the same, though it is not clearly explained why this form is used instead and might be confusing for someone comparing the implementation and the paper. I assume this is more numerically stable or something like that and that is why it is used? Maybe we should just add a comment about it?

> Within my thesis i also implemented the `Automating Entropy Adjustment` initially presented in:
> 
> > Haarnoja, T., Zhou, A., Hartikainen, K., Tucker, G., Ha, S., Tan, J., Kumar, V., Zhu, H., Gupta, A., Abbeel, P., & Levine, S. (2019). Soft Actor-Critic Algorithms and Applications. ArXiv:1812.05905 [Cs, Stat]. http://arxiv.org/abs/1812.05905
> 
> But i haven't found the time yet to adjust the implementation within this repo. Im willing to share that code if you have time to adjust the implementation.

Would be nice, but I think for now I will start with this. Might come back to you about it.

> If you want to link a paper i would recommend to link the later one. The latest development state can be tracked in the authors official repository: https://github.com/rail-berkeley/softlearning (They use tensorflow)

Sounds good.
